### PR TITLE
[doc] example and hints about pairwise distance/similarity matrices. Closes: #48306

### LINKED ISSUE #48306: [doc] example for pairwise distance matrix
## 📚 Documentation

In deep metric learning we usually have to compute a pairwise similarity/distance matrix. For example, the cosine distance matrix `pdist` is computed as:
```python
x = th.rand(10, 128)  # a batch of 128-dim embedding vectors for 10 samples
x = th.nn.functional.normalize(x, p=2, dim=-1)
pdist = 1 - th.mm(x, x.t())
```

The documentation of `th.nn.functional.cosine_similarity` looks like that it only supports a one-to-one similarity computation, namely it computes `[ cosine_similarity(x, y) for (x, y) in zip(list_x, list_y)]`. This does not look always useful.

However, by chance I find `th.nn.functional.cosine_similarity` and `th.nn.functional.pairwise_distance` able to compute the (all-to-all) distance/similarity matrix, besides the (one-to-one) distance/similarity vector:
```python
In [27]: x = th.rand(10, 128)                                                                                                          

In [28]: th.nn.functional.cosine_similarity(x[:,:,None], x.t()[None,:,:])                                                              
Out[28]: 
tensor([[1.0000, 0.7561, 0.7765, 0.7805, 0.7583, 0.7471, 0.7433, 0.7379, 0.7750,
         0.7806],
        [0.7561, 1.0000, 0.7879, 0.7083, 0.7799, 0.7824, 0.7823, 0.7193, 0.7696,
         0.7452],
        [0.7765, 0.7879, 1.0000, 0.7587, 0.7712, 0.7632, 0.7462, 0.7953, 0.7507,
         0.7367],
        [0.7805, 0.7083, 0.7587, 1.0000, 0.7618, 0.7151, 0.7183, 0.7775, 0.7471,
         0.7393],
        [0.7583, 0.7799, 0.7712, 0.7618, 1.0000, 0.7719, 0.7629, 0.7795, 0.7379,
         0.7549],
        [0.7471, 0.7824, 0.7632, 0.7151, 0.7719, 1.0000, 0.7598, 0.7502, 0.6889,
         0.6796],
        [0.7433, 0.7823, 0.7462, 0.7183, 0.7629, 0.7598, 1.0000, 0.7318, 0.7658,
         0.7437],
        [0.7379, 0.7193, 0.7953, 0.7775, 0.7795, 0.7502, 0.7318, 1.0000, 0.7357,
         0.7186],
        [0.7750, 0.7696, 0.7507, 0.7471, 0.7379, 0.6889, 0.7658, 0.7357, 1.0000,
         0.7563],
        [0.7806, 0.7452, 0.7367, 0.7393, 0.7549, 0.6796, 0.7437, 0.7186, 0.7563,
         1.0000]])

In [29]: th.nn.functional.pairwise_distance(x[:,:,None], x.t()[None,:,:])                                                              
Out[29]: 
tensor([[1.1314e-05, 4.5799e+00, 4.4812e+00, 4.3319e+00, 4.5615e+00, 4.6469e+00,
         4.6721e+00, 4.7185e+00, 4.3587e+00, 4.1595e+00],
        [4.5799e+00, 1.1314e-05, 4.3764e+00, 5.0094e+00, 4.3655e+00, 4.3234e+00,
         4.3160e+00, 4.8984e+00, 4.4245e+00, 4.4941e+00],
        [4.4812e+00, 4.3764e+00, 1.1314e-05, 4.6561e+00, 4.5461e+00, 4.6100e+00,
         4.7638e+00, 4.2785e+00, 4.7076e+00, 4.7013e+00],
        [4.3319e+00, 5.0094e+00, 4.6561e+00, 1.1314e-05, 4.5287e+00, 4.9323e+00,
         4.8955e+00, 4.3480e+00, 4.6218e+00, 4.5260e+00],
        [4.5615e+00, 4.3655e+00, 4.5461e+00, 4.5287e+00, 1.1314e-05, 4.4280e+00,
         4.5057e+00, 4.3432e+00, 4.7210e+00, 4.4123e+00],
        [4.6469e+00, 4.3234e+00, 4.6100e+00, 4.9323e+00, 4.4280e+00, 1.1314e-05,
         4.5161e+00, 4.6035e+00, 5.1214e+00, 5.0031e+00],
        [4.6721e+00, 4.3160e+00, 4.7638e+00, 4.8955e+00, 4.5057e+00, 4.5161e+00,
         1.1314e-05, 4.7602e+00, 4.4341e+00, 4.4714e+00],
        [4.7185e+00, 4.8984e+00, 4.2785e+00, 4.3480e+00, 4.3432e+00, 4.6035e+00,
         4.7602e+00, 1.1314e-05, 4.7084e+00, 4.6772e+00],
        [4.3587e+00, 4.4245e+00, 4.7076e+00, 4.6218e+00, 4.7210e+00, 5.1214e+00,
         4.4341e+00, 4.7084e+00, 1.1314e-05, 4.3396e+00],
        [4.1595e+00, 4.4940e+00, 4.7013e+00, 4.5260e+00, 4.4123e+00, 5.0031e+00,
         4.4714e+00, 4.6772e+00, 4.3396e+00, 1.1314e-05]])
```

Why isn't such usage not documented? I think it is very important for metric learning users.

### PR DESCRIPTION
**Pull request by cdluminate**
Fixes #48306

### CODE REVIEW

File: torch/nn/modules/distance.py [hunk 1]
```diff
@@ -11,6 +11,8 @@ class PairwiseDistance(Module):
     .. math ::
         \Vert x \Vert _p = \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}.
 
+    See also `torch.pdist` and `torch.cdist` for pairwise distance matrices.
```
Comment (zou3519) [inline torch/nn/modules/distance.py:14]: Cross reference should be 
```
:func:`torch.pdist`
```

**Review by zou3519 — APPROVED** [review]

**Review by zou3519 — DISMISSED** [review]
Please correct the cross-reference. Other than that, this lgtm.

**Comment by cdluminate** [comment]
@zou3519 Thanks! Revised according to the suggestions. Plus, I also added one more line of example for th.nn.CosineSimilarity
```diff
@@ -61,6 +63,7 @@ class CosineSimilarity(Module):
         >>> input2 = torch.randn(100, 128)
         >>> cos = nn.CosineSimilarity(dim=1, eps=1e-6)
         >>> output = cos(input1, input2)
+        >>> output_matrix = cos(input1[:,:,None], output2.T[None,:,:])
     """
     __constants__ = ['dim', 'eps']
     dim: int

```

**Comment by vadimkantorov** [comment]
Does this lead to materialization of the big broadcasted matrix (before the reduction is done)? On both cases, it should be mentioned in the docs

**Comment by zou3519** [comment]
@vadimkantorov, unfortunately yes it looks like: https://github.com/pytorch/pytorch/blob/eb8768651152f1107b22c6c0a891ed2ff4b40485/aten/src/ATen/native/Distance.cpp#L188

@cdluminate thank you for the update! Could you rebase this PR (there seems to be a merge conflict) and incorporate @vadimkantorov's suggestion please? Other than those two this should be good to go

**Comment by vadimkantorov** [comment]
> @vadimkantorov's suggestion please? Other than those two this should be good to go

Suggestion: explicitly mention this in docs, since it's an important thing for large-batch, large-dim metric learning. And also maybe metion alternative implementations with F.normalize/mm which do contraction and do not materialize intermediate result? Also maybe mention KeOps? or even better - bring some KeOps APIs in core :) at least the one that extends einsum into more flexible mini-DSL for contractions, allowing to take elementwise operations and custom contractions.
